source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+- The same applies to reporting the data in VM stats, unless we are willing to send the data, even when unmodified, on each VM stats query.
+- Data read after the VM stops can be considered stable and reported directly.  However, if it is attached to VM.destroy response and the response fails for any reason, it can't be reported again in a response to a repeated VM.destroy call, because the data is already deleted after the libvirt domain has been undefined.
+- A specific API call would solve most of the problems.  Vdsm could include TPM data into the computation of the VM device hash sent within VM stats and Engine could request TPM data itself if the device hash changes.  Alternatively, Engine could make the API call periodically, providing TPM data hashes as an argument, and Vdsm would respond only with data not matching the hashes.  An additional call would happen in either case just before calling VM.destroy to get the ultimate final version of TPM data.
+- It is fully dependent on Engine cooperation, it's not possible to start a VM without Engine explicitly providing the data.  But if we needed just to restart a VM on the same host without Engine then we could retain TPM data by either not undefining the domain XML or by making a local copy of the data between the restarts.
+
+This approach should be feasible with introducing a new Vdsm API call as outlined above but it has the drawback that Engine assistance is needed.
+
+#### Storing the data in domain XML metadata
+
+Using domain XML metadata, it's possible to transfer TPM data in all the directions.  TPM data can be read from the data on VM start, with or without Engine assistance, updates can be stored there atomically and Engine can update the data in monitoring.
+
+Ensuring data consistency read from swtpm is the same as with API calls.
+
+The only problem with using metadata is that it increases domain XML size significantly.  TPM data shouldn't be large although to know the size limits would require further investigation.  But initial swtpm data for a VM takes about 4 KB encoded.  It would "pollute" the domain XML and would have to be transferred each time Engine asks for the domain XML, even when there are no changes to data, which means it would be waste most of the time.
+
+This approach is a bit simpler and more flexible than using API calls but has the data size drawback.
+
+#### Using shared storage
+
+A somewhat similar approach to using domain XML metadata would be to store the data to a shared storage.  Instead of writing to or reading it from the metadata, it would be written to or read from the storage.  The benefit would be avoiding domain XML pollution and unnecessary data transfers.  Additionally, Engine wouldn't have to process the data and maintain it in the database and the data could be updated even when Engine was temporarily unavailable.
+
+However, there are also drawbacks:
+
+- A volume for TPM data must be created, arranged and maintained properly.
```

Comment:
Another drawback is that the size of these volume would likely be at lease 1G on block storage (like volumes that store vm configuration for snapshots with memory). This appears to be a reoccurring issue that also came up in the context of Q35 features..

Reply:
No, we can create 128m disks.

Reply:
Ack, so it seems that we converge here to not store the TPM data on the storage domain but it could be useful to create smaller disks for snapshot's metadata volumes - allocating 1g as we do today for xml and some other vdsm properties is a waste